File: slist.f
\ List of strings
\
\ A string is a cell holding the length in bytes, followed by the contents.
\ A list element is a string followed by a cell holding the link to the next element,
\ where 0 indicates end of list.

: contents      dup cell+  swap @ ;
: >contents     2dup !  cell+ swap move ;
: link          dup @ cell + + ;
: >link         link ! ;
: elem          here dup >r  over cell+ allot  >contents 0 , r> ; 
: .elem         dup . ." : "  dup contents type ."  -> " link @ . ;
: .list         begin dup while dup .elem cr link @ repeat drop ;
: >tail         begin link dup @ while @ repeat ! ;

: test          s" first" elem >r s" second" elem r@ >tail
                s" third" elem r@ >tail r> ;
test .list bye

File: slist.c
/* List of strings */

#include <stdio.h>
#include <stdlib.h>

typedef struct SList SList;
struct SList {
	unsigned long	len;
	char			*str;
	SList			*link;
};

SList*
elem(char *s, unsigned long len)
{
	SList *lp;

	lp = malloc(sizeof(SList));
	lp->str = s;
	lp->len = len;
	lp->link = NULL;
	return lp;
}

void
pelem(SList *lp)
{
	printf("%lu : %.*s -> %lu", (unsigned long)lp, (int)lp->len, lp->str, (unsigned long)lp->link);
}

/* Print list */
void
plist(SList *lp)
{
	SList *p;

	for(p = lp; p; p = p->link){
		pelem(p);
		putchar('\n');
	}
}

void
addtail(SList *lp, SList *ep)
{
	SList *p;

	for(p = lp; p->link; p = p->link)
		;
	p->link = ep;
}

SList*
test(void)
{
	SList *lp;

	lp = elem("first", 5);
	addtail(lp, elem("second", 6));
	addtail(lp, elem("third", 5));
	return lp;
}

int
main(void)
{
	plist(test());
	return 0;
}

	
